docs: ADR for course data independence from libraries (#33953)
openedx/edx-platform: openedx/core/djangoapps/content_libraries/docs/decisions/0003-course-content-independence.rst

3. Course Content should not depend on Library Content

> ------------------------------------------------------

Reviewer: could you update this heading to `#####` and the others headings to `*****`? https://docs.openedx.org/en/latest/documentors/references/quick_reference_rst.html#headings

Author: Sure. I was just copying the other ADRs in that folder. I'll update them as well in a separate commit.

Status
------

Accepted

Context
-------

> When a course uses content from a library via a ``LibraryContentBlock``, it creates a course-local copy of that content. It will also store certain metadata in the ModuleStore that connect the course copy with the original in the library, such as the usage key and version of the original, as well as the library-suggested values for various fields (e.g. ``num_attempts=10``).

Reviewer: ```suggestion
When a course uses content from a library via a ``LibraryContentBlock``, it creates a course-local copy of the library's blocks. For each library block, it will also store certain metadata in the ModuleStore that connects the course copy with the original in the library, such as the usage key and version of the original, as well as the library-suggested values for various fields (e.g. ``num_attempts=10``).
```

on my first read-through I took "content" to mean "a set of blocks" rather than "each block". this might clarify it a bit.

This metadata is not currently a part of the course export, and must be reconstructed by querying the library during the course import process. This causes a number of issues:

> #. The course might be imported into a site where the library does not exist, meaning that there's no way for us to figure out what the library-suggested values should be.

Reviewer: ```suggestion
#. The course might be imported into a site where the library does not exist, meaning that there's no way for us to figure out what the library-suggested values should be. This becomes a problem for any block which does not override its library-suggested value: the block will fall back to the block-type-suggested value, which is often not appropriate.
```

a bit more color on this particular problem

#. Studio Clipboard Copy & Paste functionality depends on serialization to OLX, so this metadata information is lost when an author tries to make copies of the content in other places.

We have experienced a number of `serious bugs <https://openedx.atlassian.net/wiki/spaces/COMM/pages/3858661405/Bugs+from+Content+Libraries+V1#Issues-with-update_children>`_ because of this auto-sync behavior.

Decision
--------

Content that is stored in a Course should only have to query content in a Library in the following unavoidable circumstances:

#. Initially picking the content that will be used (copied into) the Course.
#. Checking to see if an updated version of the content is available.
#. Updating the content copied into the Course with a more recent version stored in the Library (only done when a course author explicitly triggers it).

In all other cases, the course copy of the content should be self-sufficient and should not require "syncing" to a library to fill in missing metadata. In particular, the following operations should be possible to do without querying the library:

> * Import and Export of courses.

Reviewer: ```suggestion
* Import and Export of courses within and across sites.
```

optional suggestion

* Copy/Paste/Duplicate of content within and across courses.

Consequences
------------

> * We will need to serialize more metadata into the course export, probably by adding new fields to OLX.
> * This should remove a major source of bugs and confusion during the import process.

Reviewer: worth adding some more detail here?

```suggestion

Technical:

* We will need to serialize more metadata into the course export, probably by adding new fields to OLX. We will need to make this metadata available for library content operations in the XBlock runtime (ADRs to follow).
* We will need to update Import/Copy-Paste/Duplicate to use this metadata rather triggering a library sync.

User-facing:

* This should remove a major source of bugs and confusion during the import process.
* We should be able to begin preserving edits to library content during the Duplicate and Copy-Paste operations, which has been impossible as long as those operations have involved a library sync.
```

Reviewer: (EDIT: I combined the first and second technical bullet points)

3. Course Content should not depend on Library Content
######################################################

Status
******

Accepted

Context
*******

> When a course uses content from a library via a ``LibraryContentBlock``, it creates a course-local copy of the library's blocks. For each library block, it will also store certain metadata in the ModuleStore that connects the course copy with the original in the library, such as the usage key and version of the original, as well as the library-suggested values for various fields (e.g. ``num_attempts=10``).

Reviewer 2: This describes the situation with modulestore. Would it be any different with Learning Core? (I had hoped that we'd be able to do less copying and more referencing, for example.)

Author: I'm not sure. I initially hated the fact that we were copying so much instead of referencing, but I've come to appreciate the decoupling this actually gave when considering something like the v1 -> v2 upgrade process, because it would mean that old content that was never going to be used again didn't need to be upgraded. I think that the use cases we have for cross-instance import/export also point to an area where copying is the better solution.

Internally, I think it makes sense to have a schema that allows for references + copies where we cross ownership/lifecycle boundaries. But those boundaries are really fuzzy to me at the moment. I'm strongly against having a 1:1 straight mapping of v1 -> v2 libraries because I think it's a terrible missed opportunity. But if not that, then what? Manually created v2 libraries that import multiple v1 libraries? An implicit, org-level shared content library maybe? I dunno.

We have experienced a number of `serious bugs <https://openedx.atlassian.net/wiki/spaces/COMM/pages/3858661405/Bugs+from+Content+Libraries+V1#Issues-with-update_children>`_ because of this auto-sync behavior.

Decision
********

Content that is stored in a Course should only have to query content in a Library in the following unavoidable circumstances:

#. Initially picking the content that will be used (copied into) the Course.
#. Checking to see if an updated version of the content is available.
#. Updating the content copied into the Course with a more recent version stored in the Library (only done when a course author explicitly triggers it).

In all other cases, the course copy of the content should be self-sufficient and should not require "syncing" to a library to fill in missing metadata. In particular, the following operations should be possible to do without querying the library:

* Import and Export of courses within and across sites.
* Copy/Paste/Duplicate of content within and across courses.

Consequences
************

Technical:

> * We will need to serialize more metadata into the course export, probably by adding new fields to OLX. We will need to make this metadata available for library content operations in the XBlock runtime (ADRs to follow).

Reviewer 2: I think we're already doing this for copy-paste, so you could do something similar.

If you copy block A then paste it as block B, then copy block B and view the context of the clipboard OLX (`/admin/content_staging/` in the django admin), you'll see that B's OLX includes something like `copied_from_block="block-v1:BradenX+TS100+23+type@problem+block@bed1cd95990a4b6b8f327494f9fe8c77"`

Author: 😄 Yup, I remember–I was the one who asked for that feature in the first place, so that we could better track usage in analytics later on. @kdmccormick's going to put out an ADR detailing the specifics, but we're looking at a couple of new fields in this style, and maybe a `default-{field-name}` convention for handling library default values.
